hudi-common/src/main/java/org/apache/hudi/common/table/log/AbstractHoodieLogRecordReader.java

```diff
@@ -539,6 +542,121 @@ private void updateBlockSequenceTracker(HoodieLogBlock logBlock, String instantT
+            processCommandBlock(logBlock, instantToBlocksMap, targetRollbackInstants);
+            break;
+          default:
+            throw new UnsupportedOperationException("Block type not yet supported.");
+        }
+      }
+      // merge the last read block when all the blocks are done reading
+      if (!currentInstantLogBlocks.isEmpty() && !skipProcessingBlocks) {
+        LOG.info("Merging the final data blocks");
+        processQueuedBlocksForInstant(currentInstantLogBlocks, scannedLogFiles.size(), keySpecOption);
+      }
+    } catch (IOException e) {
+      LOG.error("Got IOException when reading log file", e);
+      throw new HoodieIOException("IOException when reading log file ", e);
+    } catch (Exception e) {
+      LOG.error("Got exception when reading log file", e);
+      throw new HoodieException("Exception when reading log file ", e);
+    }
+  }
+
+  private boolean shouldSkipBlock(HoodieLogBlock logBlock) {
+    if (logBlock.getBlockType().equals(CORRUPT_BLOCK)) {
+      totalCorruptBlocks.incrementAndGet();
+      return true;
+    }
+    HoodieTimeline commitsTimeline = this.hoodieTableMetaClient.getCommitsTimeline();
+    HoodieTimeline completedInstantsTimeline = commitsTimeline.filterCompletedInstants();
+    HoodieTimeline inflightInstantsTimeline = commitsTimeline.filterInflights();
+    String instantTime = logBlock.getLogBlockHeader().get(INSTANT_TIME);
+    if (logBlock.getBlockType() != COMMAND_BLOCK) {
+      if (!completedInstantsTimeline.containsOrBeforeTimelineStarts(instantTime)
+          || inflightInstantsTimeline.containsInstant(instantTime)) {
+        // hit an uncommitted block possibly from a failed write, move to the next one and skip processing this one
+        return true;
+      }
+      if (instantRange.isPresent() && !instantRange.get().isInRange(instantTime)) {
+        // filter the log block by instant range
+        return true;
+      }
+    }
+    return false;
+  }
+
+  private void processRegularDataBlock(HoodieLogBlock logBlock, ConcurrentMap<String, List<HoodieLogBlock>> instantToBlocksMap, ConcurrentMap<String, String> blockTimeToCompactionBlockTimeMap,
+                                       Deque<HoodieLogBlock> processingDeque) {
+    String instantTime = logBlock.getLogBlockHeader().get(INSTANT_TIME);
+    instantToBlocksMap.compute(instantTime, (key, val) -> {
+      if (val == null) {
+        val = new ArrayList<>();
+      }
+      val.add(logBlock);
+      return val;
+    });
+
+    // For compacted blocks
+    if (logBlock.getLogBlockHeader().containsKey(COMPACTED_BLOCK_TIMES)) {
+      Arrays.stream(logBlock.getLogBlockHeader().get(COMPACTED_BLOCK_TIMES).split(","))
+          .forEach(originalInstant -> {
+            String finalInstant = blockTimeToCompactionBlockTimeMap.getOrDefault(instantTime, instantTime);
+            blockTimeToCompactionBlockTimeMap.put(originalInstant, finalInstant);
+            if (!processingDeque.contains(logBlock)) {
```

Comment:
how large is the `processingDeque`? if it is larger than a couple elements, it may make sense to use a marker to note that the logBlock was already added as you iterate over each `originalInstant` here